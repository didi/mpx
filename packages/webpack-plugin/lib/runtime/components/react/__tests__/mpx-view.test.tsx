import React from 'react'
import { render, screen } from '@testing-library/react-native'
import View from '../mpx-view'
import Text from '../mpx-text'

// Mock mpx-portal
jest.mock('../mpx-portal', () => {
  const mockReact = require('react')
  return {
    __esModule: true,
    default: mockReact.forwardRef((props: any, ref: any) => {
      return mockReact.createElement('View', { ...props, ref })
    })
  }
})

describe('MpxView', () => {
  it('should render basic view', () => {
    const { toJSON } = render(
      <View testID="basic-view">
        <Text>Basic View Content</Text>
      </View>
    )
    
    expect(screen.getByTestId('basic-view')).toBeTruthy()
    expect(screen.getByText('Basic View Content')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should apply custom styles', () => {
    const customStyle = { 
      backgroundColor: '#f0f0f0',
      padding: 20,
      flex: 1
    }
    
    const { toJSON } = render(
      <View style={customStyle} testID="styled-view">
        <Text>Styled Content</Text>
      </View>
    )
    
    const viewElement = screen.getByTestId('styled-view')
    expect(viewElement).toBeTruthy()
    expect(viewElement.props.style).toMatchObject(customStyle)
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle nested views', () => {
    const { toJSON } = render(
      <View testID="parent-view">
        <View testID="child-view-1">
          <Text>Child 1</Text>
        </View>
        <View testID="child-view-2">
          <Text>Child 2</Text>
        </View>
      </View>
    )
    
    expect(screen.getByTestId('parent-view')).toBeTruthy()
    expect(screen.getByTestId('child-view-1')).toBeTruthy()
    expect(screen.getByTestId('child-view-2')).toBeTruthy()
    expect(screen.getByText('Child 1')).toBeTruthy()
    expect(screen.getByText('Child 2')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle accessibility props', () => {
    render(
      <View 
        testID="accessible-view"
        accessibilityLabel="Main container"
        accessibilityHint="Contains important content"
        accessibilityRole="main"
      >
        <Text>Accessible Content</Text>
      </View>
    )
    
    const viewElement = screen.getByTestId('accessible-view')
    expect(viewElement).toBeTruthy()
    expect(viewElement.props.accessibilityLabel).toBe('Main container')
    expect(viewElement.props.accessibilityHint).toBe('Contains important content')
    expect(viewElement.props.accessibilityRole).toBe('main')
  })

  it('should handle complex nested structure', () => {
    const { toJSON } = render(
      <View testID="complex-structure" style={{ flex: 1 }}>
        <View testID="header" style={{ height: 50 }}>
          <Text>Header Content</Text>
        </View>
        <View testID="body" style={{ flex: 1 }}>
          <View testID="section-1">
            <Text>Section 1</Text>
          </View>
          <View testID="section-2">
            <Text>Section 2</Text>
          </View>
        </View>
        <View testID="footer" style={{ height: 50 }}>
          <Text>Footer Content</Text>
        </View>
      </View>
    )
    
    // 验证所有组件都能被找到
    expect(screen.getByTestId('complex-structure')).toBeTruthy()
    expect(screen.getByTestId('header')).toBeTruthy()
    expect(screen.getByTestId('body')).toBeTruthy()
    expect(screen.getByTestId('section-1')).toBeTruthy()
    expect(screen.getByTestId('section-2')).toBeTruthy()
    expect(screen.getByTestId('footer')).toBeTruthy()
    
    // 验证文本内容
    expect(screen.getByText('Header Content')).toBeTruthy()
    expect(screen.getByText('Section 1')).toBeTruthy()
    expect(screen.getByText('Section 2')).toBeTruthy()
    expect(screen.getByText('Footer Content')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle empty view', () => {
    render(<View testID="empty-view"></View>)
    
    const viewElement = screen.getByTestId('empty-view')
    expect(viewElement).toBeTruthy()
    expect(viewElement.children).toEqual([])
  })

  it('should handle view with single child', () => {
    render(
      <View testID="single-child-view">
        <Text>Only Child</Text>
      </View>
    )
    
    expect(screen.getByTestId('single-child-view')).toBeTruthy()
    expect(screen.getByText('Only Child')).toBeTruthy()
  })

  it('should handle view with multiple text children', () => {
    render(
      <View testID="multi-text-view">
        <Text>First Text</Text>
        <Text>Second Text</Text>
        <Text>Third Text</Text>
      </View>
    )
    
    expect(screen.getByTestId('multi-text-view')).toBeTruthy()
    expect(screen.getByText('First Text')).toBeTruthy()
    expect(screen.getByText('Second Text')).toBeTruthy()
    expect(screen.getByText('Third Text')).toBeTruthy()
  })
})
