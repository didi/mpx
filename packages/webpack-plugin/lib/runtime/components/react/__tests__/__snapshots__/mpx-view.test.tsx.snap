// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MpxView should apply custom styles 1`] = `
<View
  style={
    {
      "backgroundColor": "#f0f0f0",
      "flex": 1,
      "padding": 20,
    }
  }
  testID="styled-view"
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={{}}
  >
    Styled Content
  </Text>
</View>
`;

exports[`MpxView should handle complex nested structure 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
  testID="complex-structure"
>
  <View
    style={
      {
        "height": 50,
      }
    }
    testID="header"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={{}}
    >
      Header Content
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
    testID="body"
  >
    <View
      style={{}}
      testID="section-1"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={{}}
      >
        Section 1
      </Text>
    </View>
    <View
      style={{}}
      testID="section-2"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={{}}
      >
        Section 2
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 50,
      }
    }
    testID="footer"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={{}}
    >
      Footer Content
    </Text>
  </View>
</View>
`;

exports[`MpxView should handle nested views 1`] = `
<View
  style={{}}
  testID="parent-view"
>
  <View
    style={{}}
    testID="child-view-1"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={{}}
    >
      Child 1
    </Text>
  </View>
  <View
    style={{}}
    testID="child-view-2"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={{}}
    >
      Child 2
    </Text>
  </View>
</View>
`;

exports[`MpxView should render basic view 1`] = `
<View
  style={{}}
  testID="basic-view"
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={{}}
  >
    Basic View Content
  </Text>
</View>
`;
